### PR TITLE
Add initial coalton file parser

### PR DESCRIPTION
**Pull request by eliaslfox**
This is the first pr for supporting Coalton code in external files. It defines a new parser subsystem and AST which will eventually replace the existing ones. The new parser uses [Concrete-Syntax-Tree](https://github.com/s-expressionists/Concrete-Syntax-Tree) and [Eclector](https://github.com/s-expressionists/Eclector) to read s-expressions with source information. 

@colescott developed a new error system based on Rust's which can generate much better errors than were possible before. A few examples below:

When parsing a class with an invalid variable:
```
(define-class (C a))
```

New error message:

```
error: Invalid type variable
  --> /Users/ejlawsonfox/lisp/coalton/test.coalton:3:24
   |
 3 | (define-class (C a))
   |                   ^ expected keyword symbol
help: add `:` to symbol
 3 | (define-class (C :a))
   |                  --
```
Previous error message:

```
Failed to parse (C A)
invalid type class predicate
In DEFINE-CLASS for class C
```

When parsing an if without an else clause:
```
(define x (if y 5))
```
New error message:
```
error: Malformed if expression
  --> /Users/ejlawsonfox/lisp/coalton/test.coalton:3:24
   |
 3 | (define x (if y 5))
   |                   ^ expected a form
```

Previous error message falls through to sbcl:

```
Error while parsing arguments to DEFMACRO IF:
  too few elements in
    (Y 5)
  to satisfy lambda list
    (EXPR THEN ELSE):
  exactly 3 expected, but got 2
```

The new parser should parse all existing syntax, however there is still more work to be done. Macro expansion is not yet implemented. The main challenge is that macros can generate forms without source information. The current error system relies entirely on source information to generate error messages and will require further extension.

Additional work will also be needed to allow running the new parser in the coalton and coalton-toplevel macros. Macro based invocations will require either generating errors without source information or somehow getting source information from SBCL.

Test coverage for the parser is pretty good. Every branch is hit and every error message is signalled. Coverage can be generated locally with `make parser-coverage`. 

Screenshot_2022-11-01 Screenshot

Also note this this pr is into the branch coalton-file which will be used to track work on coalton files.

### CODE REVIEW

**Comment by gefjon** [comment]
Very cool!

> Macro expansion is not yet implemented. The main challenge is that macros can generate forms without source information. The current error system relies entirely on source information to generate error messages and will require further extension.

I see two ways forward for this. The version that makes writing macros easy at the cost of error messages is, when Coalton calls `macroexpand` on a form A to produce a form B, it copies A's source information into any parts of B that don't have source information already. The version that generates better error messages has been explored by Racket's syntax objects; essentially, macros are responsible for adding source information tags to their output.

> Additional work will also be needed to allow running the new parser in the coalton and coalton-toplevel macros. Macro based invocations will require either generating errors without source information or somehow getting source information from SBCL.

I am somewhat loathe to suggest this, but this seems like a classic use-case for reader macros. If we replace `(coalton ...)` and `(coalton-toplevel ...)` with, say, `#{coalton ...}` or `#{coalton-toplevel ...}`, we can parse the `...` the way we want. It's not ideal ergonomically, as it will require any Common Lisp code that wants to invoke Coalton that way to use a named-readtable, but I think it's probably better than maintaining two separate parsers which generate different error messages.

**Comment by eliaslfox** [comment]
> The version that makes writing macros easy at the cost of error messages is, when Coalton calls macroexpand on a form A to produce a form B, it copies A's source information into any parts of B that don't have source information already.

This seems quite implementable. 

> I am somewhat loathe to suggest this, but this seems like a classic use-case for reader macros

That's the current plan.

**Review by stylewarning — CHANGES_REQUESTED** [review]
some initial review

File: src/parser/base.lisp [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+  (declare (type sb-sys::fd-stream file)
+           (type integer index)
+           (values string &optional))
+  (file-position file 0)
+  (loop :for i :from 1 :to index
+        :for line := (read-line file)
+        :when (= i index)
+          :do (return-from get-nth-line line))
+
+  (util:unreachable))
```
[the remark this answers is not in the document]
Comment (eliaslfox) [reply src/parser/base.lisp]: good catch

File: src/parser/base.lisp [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+Returns (VALUES LINE-NUM LINE-START-INDEX)"
+  (declare (type sb-sys::fd-stream file)
+           (type integer index)
+           (values integer integer))
+  (file-position file 0)
+  (let ((line 1)
+        (line-start-index 0))
+    (loop :for i :to index
+          :for char := (read-char file)
+          :when (char= char #\Newline)
```
[the remark this answers is not in the document]
Comment (eliaslfox) [reply src/parser/base.lisp:42]: yes :(

File: src/parser/parser.lisp [hunk 3]
```diff
@@ -0,0 +1,1491 @@
...
+           ;; Read unspecified floats as double floats
+           (*read-default-float-format* 'double-float)
+
+           ;; Read the (package) form
+           (package-form (eclector.concrete-syntax-tree:read file nil 'eof)))
+
+      (when (eq package-form 'eof)
+        (error "file starts with eof"))
+
+      (setf *package* (parse-package package-form file))
```
Comment (stylewarning) [inline src/parser/parser.lisp:321]: is there a reason to not bind `*package*` instead of setting it for the context of this function?

File: src/parser/parser.lisp [hunk 4]
```diff
@@ -0,0 +1,1491 @@
...
+
+      (let* ((program (make-program :package *package* :file file))
+
+             (attributes (make-array 0 :adjustable t :fill-pointer t)))
+
+        (block parse-loop
+          (loop :with elem := nil
```
Comment (stylewarning) [inline src/parser/parser.lisp:328]: you can use

```
(loop :named parse-loop
```
 
instead of the `block`

File: src/parser/parser.lisp [hunk 5]
```diff
@@ -0,0 +1,1491 @@
...
+                :do (setf elem (eclector.concrete-syntax-tree:read file nil 'eof))
+
+                :when (eq elem 'eof)
+                  :do (return-from parse-loop)
```
Comment (stylewarning) [inline src/parser/parser.lisp:332]: actually, if all you're doing is returning, I think you can just do `(return)`

File: src/parser/parser.lisp [hunk 6]
```diff
@@ -0,0 +1,1491 @@
...
+
+                :do (when (and (parse-toplevel-form elem program attributes file)
+                               (plusp (length attributes)))
+                      (util:coalton-bug "parse-toplevel-form indicated that a form was parsed but did not
+consume all attributes"))))
+
+        (unless (zerop (length attributes))
```
Comment (stylewarning) [inline src/parser/parser.lisp:339]: zerop length => endp

File: src/parser/parser.lisp [hunk 7]
```diff
@@ -0,0 +1,1491 @@
...
+        program))))
+
+(defun parse-package (form file)
+  "Parses a coalton package decleration in the form of (package {name})"
+  (declare (type cst:cst form)
+           (type sb-sys::fd-stream file)
+           (values package))
+
+  ;; Package declarations must start with "PACKAGE"
+  (unless (equalp (symbol-name (cst:raw (cst:first form))) "PACKAGE")
```
Comment (stylewarning) [inline src/parser/parser.lisp:360]: `string=`

File: src/parser/parser.lisp [hunk 8]
```diff
@@ -0,0 +1,1491 @@
...
+  ;; Toplevel forms must begin with an atom
+  (when (cst:consp (cst:first form))
+    (error 'parse-error
+           :err (coalton-error
+                 (cst:first form) file
+                 :message "Malformed toplevel form"
+                 :primary-note "unexpected list")))
+
+  (case (cst:raw (cst:first form))
+    (coalton:monomorphize
```
Comment (stylewarning) [inline src/parser/parser.lisp:416]: again I might recommend `((coalton:monomorphize) ...` for cases

File: src/parser/parser.lisp [hunk 9]
```diff
@@ -0,0 +1,1491 @@
...
+
+          (setf (fill-pointer attributes) 0)
+          (setf (toplevel-define-monomorphize define) monomorphize)
+          (push define (program-defines program))
+          (return-from parse-toplevel-form t)))
+
+    (coalton:declare
+     (let ((declare (parse-declare form file))
+
+           (monomorphize)
```
Comment (stylewarning) [inline src/parser/parser.lisp:483]: no parens

File: src/parser/parser.lisp [hunk 10]
```diff
@@ -0,0 +1,1491 @@
...
+                    (setf monomorphize attribute)
+                    (setf monomorphize-form attribute-form))))
+
+       (setf (fill-pointer attributes) 0)
+       (setf (toplevel-declare-monomorphize declare) monomorphize)
+       (push declare (program-declares program))
+       (return-from parse-toplevel-form t)))
+
+    (coalton:define-type
+        (let* ((type (parse-define-type form file))
```
Comment (stylewarning) [inline src/parser/parser.lisp:528]: just `let`, and uninitialized vars parens

File: src/parser/parser.lisp [hunk 11]
```diff
@@ -0,0 +1,1491 @@
...
+
+          (setf (fill-pointer attributes) 0)
+          (setf (toplevel-define-type-repr type) repr)
+          (push type (program-types program))
+          (return-from parse-toplevel-form t)))
+
+    (coalton:define-class
+        (let ((class (parse-define-class form file)))
+
+          (unless (zerop (length attributes))
```
Comment (stylewarning) [inline src/parser/parser.lisp:577]: endp

File: src/parser/parser.lisp [hunk 12]
```diff
@@ -0,0 +1,1491 @@
...
+                           :span (toplevel-define-class-head-src class)
+                           :message "while parsing define-class")))))
+
+          (push class (program-classes program))
+          (return-from parse-toplevel-form t)))
+
+    (coalton:define-instance
+        (let ((instance (parse-define-instance form file)))
+
+          (unless (zerop (length attributes))
```
Comment (stylewarning) [inline src/parser/parser.lisp:596]: endp

File: src/parser/parser.lisp [hunk 13]
```diff
@@ -0,0 +1,1491 @@
...
+
+
+(defun parse-define (form file)
+  (declare (type cst:cst form)
+           (type sb-sys::fd-stream file)
+           (values toplevel-define))
+
+  (assert (cst:consp form))
+
+  (let ((name)
```
Comment (stylewarning) [inline src/parser/parser.lisp:628]: push the let down and possibly use multiple-value-bind

File: src/parser/parser.lisp [hunk 14]
```diff
@@ -0,0 +1,1491 @@
...
+                  :message "Invalid toplevel form"
+                  :primary-note "unknown toplevel form")))))
+
+
+(defun parse-define (form file)
+  (declare (type cst:cst form)
+           (type sb-sys::fd-stream file)
+           (values toplevel-define))
+
+  (assert (cst:consp form))
```
Comment (stylewarning) [inline src/parser/parser.lisp:626]: does this need to change?

File: src/parser/parser.lisp [hunk 15]
```diff
@@ -0,0 +1,1491 @@
...
+     :body body
+     :monomorphize nil
+     :source (cst:source form))))
+
+(defun parse-declare (form file)
+  (declare (type cst:cst form)
+           (type sb-sys::fd-stream file)
+           (values toplevel-declare))
+
+  (assert (cst:consp form))
```
Comment (stylewarning) [inline src/parser/parser.lisp:665]: does this need to change?

File: src/parser/parser.lisp [hunk 16]
```diff
@@ -0,0 +1,1491 @@
...
+
+    ;; (define-type)
+    (unless (cst:consp (cst:rest form))
+      (error 'parse-error
+             :err (coalton-error
+                   form file
+                   :message "Malformed type definition"
+                   :primary-note "expected body")))
+
+    (if (cst:atom (cst:second form))
```
Comment (stylewarning) [inline src/parser/parser.lisp:729]: prefer cond over if for stateful actions

File: src/parser/parser.lisp [hunk 17]
```diff
@@ -0,0 +1,1491 @@
...
+     :source (cst:source form))))
+
+(defun parse-define-class (form file)
+  (declare (type cst:cst form)
+           (type sb-sys::fd-stream file)
+           (values toplevel-define-class))
+
+  (assert (cst:consp form))
+
+  (let ((unparsed-name)
```
Comment (stylewarning) [inline src/parser/parser.lisp:807]: push lower

File: src/parser/parser.lisp [hunk 18]
```diff
@@ -0,0 +1,1491 @@
...
+
+      ;; (... =>)
+      (when (and left right (null (cdr right)))
+        (error 'parse-error
+               :err (coalton-error
+                     (cst:second form) file
+                     :message "Malformed class definition"
+                     :primary-note "missing class name")))
+
+      (if (null right)
```
Comment (stylewarning) [inline src/parser/parser.lisp:892]: cond

File: src/parser/parser.lisp [hunk 19]
```diff
@@ -0,0 +1,1491 @@
...
+        :head-src (cst:source (cst:second form))))))
+
+(defun parse-define-instance (form file)
+  (declare (type cst:cst form)
+           (type sb-sys::fd-stream file)
+           (values toplevel-define-instance))
+
+  (assert (cst:consp form))
+
+  (let ((unparsed-context)
```
Comment (stylewarning) [inline src/parser/parser.lisp:991]: push let down

File: src/parser/parser.lisp [hunk 20]
```diff
@@ -0,0 +1,1491 @@
...
+           (values (or null string) node-body))
+
+  (let ((docstring)
+        (unparsed-body))
+
+    ;; (define y 2)
+    (when (cst:atom (cst:rest form))
+      (return-from parse-definition-body (values nil (parse-body form enclosing-form file))))
+
+    (if (and (cst:atom (cst:first form))
```
Comment (stylewarning) [inline src/parser/parser.lisp:1298]: cond

File: src/parser/parser.lisp [hunk 21]
```diff
@@ -0,0 +1,1491 @@
...
+
+    (values docstring (parse-body unparsed-body enclosing-form file))))
+
+(defun parse-instance-method-definition (form parent-form file)
+  (declare (type cst:cst form)
+           (type cst:cst parent-form)
+           (type sb-sys::fd-stream file)
+           (values instance-method-definition))
+
+  (let* ((name)
```
Comment (stylewarning) [inline src/parser/parser.lisp:1314]: let* -> let; extra parens

File: src/parser/parser.lisp [hunk 22]
```diff
@@ -0,0 +1,1491 @@
...
+    (unless (cst:consp (cst:rest form))
+      (error 'parse-error
+             :err (coalton-error
+                   form file
+                   :message "Malformed method definition"
+                   :primary-note "expected definition name"
+                   :notes (list context-note))))
+
+
+    (setf (values name arguments) (parse-argument-list (cst:second form) file))
```
Comment (stylewarning) [inline src/parser/parser.lisp:1356]: i'd M-V-B

File: src/parser/parser.lisp [hunk 23]
```diff
@@ -0,0 +1,1491 @@
...
+           :err (coalton-error
+                 form file
+                 :message "Malformed functional dependency"
+                 :primary-note "unexpected dotted list")))
+
+  (multiple-value-bind (left right)
+      (util:take-until
+       (lambda (cst)
+         (and (cst:atom cst)
+              (eq (cst:raw cst) 'coalton:->)))
```
Comment (stylewarning) [inline src/parser/parser.lisp:1387]: prefer unparenthesized constant first

File: src/parser/parser.lisp [hunk 24]
```diff
@@ -0,0 +1,1491 @@
...
+       (cst:listify form))
+
+    (unless left
+      (error 'parse-error
+             :err (coalton-error
+                   form file
+                   :message "Malformed functional dependency"
+                   :primary-note "expected one or more type variables")))
+
+    (unless (cdr right)
```
Comment (stylewarning) [inline src/parser/parser.lisp:1397]: cdr -> rest?

File: src/parser/parser.lisp [hunk 25]
```diff
@@ -0,0 +1,1491 @@
...
+  (unless (cst:consp (cst:rest form))
+    (error 'parse-error
+           :err (coalton-error
+                 form file
+                 :highlight :end
+                 :message "Malformed repr attribute"
+                 :primary-note "expected keyword symbol")))
+
+  (let ((type (parse-type-variable (cst:second form) file)))
+    (if (eq (keyword-src-name type) :native)
```
Comment (stylewarning) [inline src/parser/parser.lisp:1446]: cond

File: src/parser/pattern.lisp [hunk 26]
```diff
@@ -0,0 +1,120 @@
...
+           (type sb-sys::fd-stream file))
+
+  (when (cst:atom form)
+    (when (typep (cst:raw form) 'util:literal-value)
+      (return-from parse-pattern
+        (make-pattern-literal
+         :value (cst:raw form)
+         :source (cst:source form))))
+
+    (when (eq (cst:raw form) 'coalton:_)
```
Comment (stylewarning) [inline src/parser/pattern.lisp:75]: constant first

File: src/parser/pattern.lisp [hunk 27]
```diff
@@ -0,0 +1,120 @@
...
+   #:pattern-constructor                ; STRUCT
+   #:make-pattern-constructor           ; CONSTRUCTOR
+   #:pattern-constructor-name           ; ACCESSOR
+   #:pattern-constructor-patterns       ; ACCESSOR
+   #:parse-pattern                      ; FUNCTION
+   ))
+
+(in-package #:coalton-impl/parser/pattern)
+
+(defstruct (pattern
```
Comment (stylewarning) [inline src/parser/pattern.lisp:32]: after reading a good number of these files, and now onto this one, I have a general feeling that we have a lot of code that's clearly following some sort of "informal specification" that's not written down. If I didn't know Coalton codebase somewhat (and Coalton itself), I would have a hard time navigating everything.

It doesn't help that documentation is incredibly light. There is some helpful comments in long case analyses (e.g., about errors), but there's nothing as it pertains to what anything means, what invariants need to be satisfied, etc. etc.

I don't want to block this PR over it, but we need to spec something out somewhere, whether it's in the headers of these files, or somewhere else.

File: src/parser/pattern.lisp [hunk 28]
```diff
@@ -0,0 +1,120 @@
...
+            (:include pattern)
+            (:copier nil)))
+
+(defstruct (pattern-constructor
+            (:include pattern)
+            (:copier nil))
+  (name     (util:required 'name)     :type identifier   :read-only t)
+  (patterns (util:required 'patterns) :type pattern-list :read-only t))
+
+(defun parse-pattern (form file)
```
Comment (stylewarning) [inline src/parser/pattern.lisp:64]: In reading this code, maybe which is just my personal history as a lisper, I'm getting a little mentally exhausted at the "early return"/"nested when"/"declare now set later" style of coding. I think this style of coding is a welcome sign to complex bugs, since it invites extension simply by splicing in new code in the middle of existing code. `PROGN` is the root of all evil, etc.

While I sometimes think sticking to a strict "expression-only" style of coding can lead to obtuse code (especially when you need to do a complex initialization), in this code, it's getting hard to follow what invariants are true when, since what an expression means now depends on following a sequence of statements, as opposed to reading a nested structure.

File: src/parser/pattern.lisp [hunk 29]
```diff
@@ -0,0 +1,120 @@
...
+                 :primary-note "unknown pattern literal")))
+
+  (unless (cst:proper-list-p form)
+    (error 'parse-error
+             :err (coalton-error
+                   form file
+                   :message "Invalid match branch"
+                   :primary-note "unexpected dotted list")))
+
+  (let ((name)
```
Comment (stylewarning) [inline src/parser/pattern.lisp:99]: push let down

**Review by stylewarning — CHANGES_REQUESTED** [review]
a more comprehensive review

File: src/parser/base.lisp [hunk 30]
```diff
@@ -0,0 +1,342 @@
...
+  '(and symbol (not boolean) (not keyword)))
+
+(defun identifier-p (x)
+  (typep x 'identifier))
+
+(defun get-line-from-index (file index)
+  "Get the line number corresponding to the character offset INDEX.
+
+Returns (VALUES LINE-NUM LINE-START-INDEX)"
+  (declare (type sb-sys::fd-stream file)
```
[the remark this answers is not in the document]
Comment (eliaslfox) [reply src/parser/base.lisp:34]: I can change this. Regular streams don't always support `file-position`, and the clhs doesn't define a type for "streams that do support file-position."

I think the best way to avoid SBCL specific functionality would be to run CI against another lisp implementation. I'm not sure what that lisp implementation would be though.

File: src/parser/expression.lisp [hunk 31]
```diff
@@ -0,0 +1,1209 @@
...
+   #:node-do                            ; STRUCT
+   #:make-node-do                       ; CONSTRUCTOR
+   #:node-do-nodes                      ; ACCESSOR
+   #:node-do-last-node                  ; ACCESSOR
+   #:parse-expression                   ; FUNCTION
+   #:parse-body                         ; FUNCTION
+   #:parse-variable                     ; FUNCTION
+   ))
+
+;; TODO: declare in let expressions
```
[the remark this answers is not in the document]
Comment (eliaslfox) [reply src/parser/expression.lisp:124]: This is done. I forgot the TODO.

File: src/parser/expression.lisp [hunk 32]
```diff
@@ -0,0 +1,1209 @@
...
+                 :message "Malformed expression"
+                 :primary-note "unexpected dotted list")))
+
+  ;;
+  ;; Keywords
+  ;;
+
+  (when (cst:atom (cst:first form))
+    (case (cst:raw (cst:first form))
+      (coalton:fn
```
[the remark this answers is not in the document]
Comment (eliaslfox) [reply src/parser/expression.lisp:397]: Thanks. I didn't realize that was an option.

File: src/parser/base.lisp [hunk 33]
```diff
@@ -0,0 +1,344 @@
...
+
+                        (when (/= end-line start-line)
+                          (setf contains-multiline-note t))
+
+                        (list note start-line start-offset end-line end-offset))))))
+              notes-with-line-info))
+
+           ;; Get the last line mentioned
+           (line-number-width
+             (1+ (floor (log (fourth (car (last notes-with-line-info))) 10)))))
```
[the remark this answers is not in the document]
Comment (gefjon) [reply src/parser/base.lisp:205]: this feels like it should be a named function

File: src/parser/base.lisp (hunk 30 above)
[the remark this answers is not in the document]
Comment (gefjon) [reply src/parser/base.lisp:34]: Pretty much the only option IMO is CCL unless you can convince Franz to get your hands on an Allegro CI setup.

File: src/parser/expression.lisp [hunk 34]
```diff
@@ -0,0 +1,1209 @@
...
+                   :message "when parsing do expression")))))
+
+  (parse-expression form file))
+
+(defun parse-let-declare (form file)
+  (declare (type cst:cst form)
+           (type sb-sys::fd-stream file)
+           (values node-let-declare))
+
+  (assert (cst:consp form))
```
[the remark this answers is not in the document]
Comment (eliaslfox) [reply src/parser/expression.lisp:1191]: This is correct. The assert is to indicate that the invariant was previously checked an a user friendly error would have been signaled already.

File: src/parser/parser.lisp (hunk 7 above)
Comment (gefjon) [reply src/parser/parser.lisp:360]: Or `string-equalp`, depending on if you want case-sensitivity.

File: src/parser/parser.lisp (hunk 11 above)
Comment (gefjon) [reply src/parser/parser.lisp:577]: or just `when attributes`

File: src/parser/parser.lisp (hunk 14 above)
Comment (eliaslfox) [reply src/parser/parser.lisp:626]: no

File: src/parser/parser.lisp (hunk 15 above)
Comment (eliaslfox) [reply src/parser/parser.lisp:665]: no

File: src/parser/parser.lisp (hunk 6 above)
Comment (eliaslfox) [reply src/parser/parser.lisp:339]: endp doesn't work on vectors

**Comment by gefjon** [comment]
> As for coding style, the decisions made in the parser are fairly intentional and I would prefer not to change them. Writing code in a more expression based style with cond and multiple-value-bind would make functions denser and push code further to the right. Also most predeclared variables are later assigned to struct slots with declared types which ensures their eventual correct definition.

> For the smaller stuff, let* is preferred over let because I have never wanted let when defining multiple variables. (let ((x)) ...) is preferred over (let (x) ...) because it keeps variable definitions consistent. Both of these match the style used throughout the rest of the codebase.

I agree with some of these, and the ones I disagree with I'm willing to accept as well-thought-out preference, but I think @stylewarning is right that it would be good to have a LISP-STYLE.md or something somewhere in the project that lays out the rationale and the fact that these are intentional decisions. But that should probably be a separate PR, I suppose. I'll open an issue.

**Comment by stylewarning** [comment]
> > As for coding style, the decisions made in the parser are fairly intentional and I would prefer not to change them. Writing code in a more expression based style with cond and multiple-value-bind would make functions denser and push code further to the right. Also most predeclared variables are later assigned to struct slots with declared types which ensures their eventual correct definition.
> 
> 
> 
> > For the smaller stuff, let* is preferred over let because I have never wanted let when defining multiple variables. (let ((x)) ...) is preferred over (let (x) ...) because it keeps variable definitions consistent. Both of these match the style used throughout the rest of the codebase.
> 
> 
> 
> I agree with some of these, and the ones I disagree with I'm willing to accept as well-thought-out preference, but I think @stylewarning is right that it would be good to have a LISP-STYLE.md or something somewhere in the project that lays out the rationale and the fact that these are intentional decisions. But that should probably be a separate PR, I suppose. I'll open an issue.

I will not be agreeable to large deviations from the style of Lisp written for the past many years from dozens of people within the Quil, Coalton, and other adjacent projects, which also deviate from every textbook written in Lisp, which also deviate from many open-source Lisp libraries. How Coalton compiler code has been written over the past year or so is a valid precedent to consider, but it wasn't a result of any principled discussion among developers (though it may have been principled in the heads of individual developers who wrote it).

Especially when a project is getting off the ground, I'm sympathetic to not letting stylistic concerns retard development velocity, as long as architectural concerns are, from a global point-of-view, in decent standing. Despite this, much to my chagrin, even my small one-character requests to keep coding style consistent have been insouciantly sidelined with little to offer in return as justification besides "I don't like it" or "I don't see the point".

Things are getting more serious, not the least of which because Coalton is depended upon in critical projects, and so I take a bit more of a serious stance in coming to an agreement for how we write code.

Pushing code to the right, or code density (at least in this context), are not without merit, but haven't hitherto been accompanied by anything objective or definitive about why it's better or worse than the idiomatic alternative. I offered a concrete reason: using expression-oriented programming as a matter of style, and explicitly delineating non-expression-oriented code through a series of stylistic rules, means that a programmer can more readily rely on assumptions about dependence, control flow, context, etc. when both understanding the code and making refactoring decisions. More specifically, nesting offers a syntactic notion of dependence, whereas concatenation merely implies it.

While there may be said merit in other styles of programming, there's value in following idioms, especially in Lisp code that is ultimately driven by an evolving team of people over years or even decades, *especially* in a language like Common Lisp—a language so frequently (and in my opinion, improperly) regarded as a language with too much flexibility invariably leading to everybody adopting their own incompatible pet styles.

If a `LISP-STYLE.md` is what's needed to rectify this and ensure agreed-upon consistency, then so be it; that has value of being unambiguous. With that said, I'm not convinced relatively routine code of this PR's nature warrants special treatment or a re-philosophization of Lisp style.

This note shouldn't be construed as an unquestionable demand to change something (especially if we now consider such change to be blocked by an agreed style guide), but it is a request for us to broaden our view of who stakeholders are/will be and consequently how decisions affect those people.

**Comment by eliaslfox** [comment]
Closed in favor of #810.
